astronomer_starship/compat/starship_operator.py

```diff
@@ -0,0 +1,141 @@
+TASK_INSTANCES = "/api/starship/task_instances"
+DAGS = "/api/starship/dags"
+
+logger = getLogger(__name__)
+
+
+def session_with_retry(retries=3, backoff_factor=2):
+    sess = requests.Session()
+    retry = Retry(
+        total=retries,
+        backoff_factor=backoff_factor,
+        status_forcelist=[500, 502, 503, 504],
+    )
+    sess.mount("http://", HTTPAdapter(max_retries=retry))
+    sess.mount("https://", HTTPAdapter(max_retries=retry))
+    return sess
+
+
+def _request(
+    type: Literal["get", "post", "put", "patch"],
+    endpoint,
+    auth,
+    json=None,
+    params=None,
+    retries=3,
+    backoff_factor=2,
+):
+    s = session_with_retry(retries=retries, backoff_factor=backoff_factor)
+    request_mapping = {"get": s.get, "post": s.post, "put": s.put, "patch": s.patch}
+    method = request_mapping.get(type)
+    resp = method(endpoint, params=params, json=json, auth=auth)
+    logger.info(f"request status {resp.status_code} for endpoint {endpoint}")
+    return resp
+
+
+# todo: maybe create utility classes?
+def get_dags(webserver_url, auth):
+    dags = urljoin(webserver_url, DAGS)
+    resp = _request("get", endpoint=dags, auth=auth)
+    return resp.json()
+
+
+def get_dagruns(webserver_url, dag_id, auth, limit=5) -> dict:
+    dagrun_endpoint = urljoin(webserver_url, DAG_RUNS)
+    resp = _request(
+        type="get",
+        endpoint=dagrun_endpoint,
+        auth=auth,
+        params={"dag_id": dag_id, "limit": limit},
+    )
+    return resp.json()
+
+
+def set_dagruns(webserver_url: str, auth, dag_runs: list[dict]) -> dict:
+    dagrun_endpoint = urljoin(webserver_url, DAG_RUNS)
+    resp = _request(
+        type="post", endpoint=dagrun_endpoint, auth=auth, json={"dag_runs": dag_runs}
+    )
+    return resp.json()
+
+
+def get_latest_dagrun_state(webserver_url: str, dag_id: str, auth: str) -> str:
+    latest = get_dagruns(webserver_url=webserver_url, dag_id=dag_id, auth=auth, limit=1)
+    if latest.status_code != 200:
+        raise Exception(
+            f"Retriveing latest dagrun failed with status: {latest.status_code} {latest.text}"
+        )
+
+    return latest[0]["state"]
+
+
+# another reason for class to couple dagrun and task instance retrieval limits
+def get_task_instances(
+    webserver_url: str, dag_id: str, auth: str, limit: int = 5
+) -> requests.Response:
+    task_instances = urljoin(webserver_url, TASK_INSTANCES)
+    resp = _request(
+        type="get",
+        endpoint=task_instances,
+        auth=auth,
+        params={"dag_id": dag_id, "limit": limit},
+    )
+    return resp
+
+
+def set_dag_state(
+    webserver_url: str,
+    dag_id: str,
+    auth,
+    action=Literal["pause", "unpause"],
+):
+    action_dict = {"pause": True, "unpause": False}
+    is_paused = action_dict[action]
+    payload = {"dag_id": dag_id, "is_paused": is_paused}
+    dag_endpoint = urljoin(webserver_url, DAGS)
+    return _request(type="patch", endpoint=dag_endpoint, auth=auth, json=payload)
```

Comment:
can you put the functionality in a `StarshipHook` type of class? Then the operator can be pretty minimal, and just call the hook.

looking good so far 👍 

There was some conversation on the previous operator that I remember on 
"Should there be an operator to migrate variables, and one to migrate pools, and etc etc etc"
vs 
"should there be one operator that does everything" 

I think we settled on the second, but I think we can revisit that conversation.
If you want to only focus on the DAG Runs (as that's what you need), that's okay 👍 having a solid base makes it real easy to build on top

Reply:
That makes sense! I'll do a quick refactor to put all the utils into a hook.

Separately, what would you think about creating a table in the source's metadata db to track what's been migrated? The user could then run this dag multiple times and it would be idempotent. Currently, I'm skipping migration of dags if it is actively running or queued. Open to discussing if that's a good idea. I initially implemented it this way in case customers are using xcoms, as we couldn't easily transfer those between deployments

Reply:
Making a table in the customer's deployment sounds like territory where things could go wrong.
I'd say let's build it to be one-and-done. 
Same as the Starship UI - if they need to reset, they just delete what was migrated and run it again

Reply:
The operator could take on DAG (or a list of dag_id's?) to migrate, so the user would be able to trigger it per-dag, when they know they've paused it 🤷 :
```
with DAG("starship_dag", schedule=None, params={"dag_id": "NOTSET", ...}, ...):
   StarshipOperator(..., dag_id="{{ params.dag_id }}")
```

Reply:
If we need idempotency - I believe we were also talking about upsert behavior on the POST route in general?

Reply:
> in case customers are using xcoms
yea, good point, though starship has _never_ transferred xcoms and likely won't (too complex, too much data, too many sources, probably not needed my 99% of folks)